### PR TITLE
daemon: support prefix id for cli operation

### PR DESCRIPTION
**Pull request by FeelyChau**
We can use prefix id for cli, like:
```sh
$ pipcook job list                                              
┌─────────┬────────────┬────────────┬───────────┬──────────────┬────────────────────────────┐
│ (index) │     id     │ pipelineId │  status   │ evaluatePass │         createdAt          │
├─────────┼────────────┼────────────┼───────────┼──────────────┼────────────────────────────┤
│    0    │ '5ck8g5no' │ 'he2tneg1' │ 'success' │     true     │ '2020-10-20T11:46:31.421Z' │
│    1    │ 'zwjlv9em' │ 'lx6k8m2h' │ 'success' │     true     │ '2020-10-20T11:45:44.670Z' │
│    2    │ '5ndv9z7x' │ '8mfffbnw' │ 'success' │     true     │ '2020-10-20T11:45:13.469Z' │
│    3    │ 'n54s87da' │ 'h6jpegv0' │ 'success' │     true     │ '2020-10-20T11:44:41.048Z' │
│    4    │ 'sv85mioz' │ 'pqekkfh7' │ 'success' │     true     │ '2020-10-20T11:43:38.121Z' │
│    5    │ 'azu0moag' │ 'ow97bsux' │ 'success' │     true     │ '2020-10-20T11:42:58.452Z' │
└─────────┴────────────┴────────────┴───────────┴──────────────┴────────────────────────────┘
$ pipcook job log 5
✖ multiple jobs found with prefix: 5
$ ipcook job log 5c      
ℹ ["Building prefix dict from the default dictionary ...
$ pipcook job log 5ck8g5no
ℹ ["Building prefix dict from the default dictionary ...
```

### CODE REVIEW

File: packages/daemon/src/model/plugin.ts [hunk 1]
```diff
@@ -86,6 +84,7 @@ export class PluginModel extends Model {
   }
 
   static async findOrCreateByParams(creationParameter: CreationParameter): Promise<PluginEntity> {
+    console.log('create', creationParameter);
```
Comment (yorkie) [inline packages/daemon/src/model/plugin.ts:87]: Please remove the debug log.

File: packages/daemon/src/model/plugin.ts [hunk 2]
```diff
@@ -49,24 +49,22 @@ export class PluginModel extends Model {
...
-
   static async findById(id: string): Promise<PluginEntity> {
     return (await PluginModel.findOne({ where: { id } }))?.toJSON() as PluginEntity;
   }
 
   static async findByIds(ids: string[]): Promise<PluginEntity[]> {
     return (await PluginModel.findAll({ where: { id: ids } })).map(plugin => plugin.toJSON() as PluginEntity);
   }
+
+  static async findByPrefixId(prefixId: string): Promise<PluginEntity[]> {
```
Comment (yorkie) [inline packages/daemon/src/model/plugin.ts:60]:
```suggestion
  static async searchById(keyword: string): Promise<PluginEntity[]> {
```

File: packages/daemon/src/model/plugin.ts [hunk 3]
```diff
@@ -49,24 +49,22 @@ export class PluginModel extends Model {
     return (await PluginModel.findAll({ where })).map(plugin => plugin.toJSON() as PluginEntity);
   }
 
-  static async query(filter?: ListPluginsFilter): Promise<PluginEntity[]> {
```
Comment (yorkie) [inline packages/daemon/src/model/plugin.ts:52]: Why remove this method?

File: packages/daemon/src/model/pipeline.ts [hunk 4]
```diff
@@ -59,6 +59,20 @@ export class PipelineModel extends Model {
     }))?.toJSON() as PipelineEntity;
   }
 
+  static async getPipelineByName(name: string): Promise<PipelineEntity> {
+    return (await PipelineModel.findOne({
+      where: { name }
+    }))?.toJSON() as PipelineEntity;
+  }
+
+  static async getPipelinesByPrefixId(prefixId: string): Promise<PipelineEntity[]> {
```
Comment (yorkie) [inline packages/daemon/src/model/pipeline.ts:68]: Ditto, I don't get what PrefixId is from its name.

Comment (yorkie) [reply packages/daemon/src/model/pipeline.ts:68]: In the name conventions, we will use "find**" if the result is an array.

File: packages/daemon/src/model/job.ts [hunk 5]
```diff
@@ -42,6 +42,14 @@ export class JobModel extends Model {
     })).map(job => job.toJSON() as JobEntity);
   }
 
+  static async findByPrefixId(prefixId: string): Promise<JobEntity[]> {
```
Comment (yorkie) [inline packages/daemon/src/model/job.ts:45]: Ditto.

File: packages/daemon/src/app/controller/plugin.ts [hunk 6]
```diff
@@ -77,23 +89,16 @@ export class PluginController extends BaseEventController {
    */
   @get('/:id')
   public async get() {
-    const plugin = await this.pluginManager.findById(this.ctx.params.id);
-    if (plugin) {
-      this.ctx.success(plugin);
-    } else {
-      this.ctx.throw(HttpStatus.NOT_FOUND, 'no plugin found');
-    }
+    const plugin = await this.fetchPluginByIdPrefix(this.ctx.params.id);
```
Comment (yorkie) [inline packages/daemon/src/app/controller/plugin.ts:92]: I'm -1 on this change, we would keep the id to be the string that matches the actual id in the database, why not just implementing this in CLI?

**Review by yorkie — CHANGES_REQUESTED** [review]

**Comment by yorkie** [comment]
I'm -1 on this current implementation which changes the meaning of id in most APIs and it is incorrect.

File: packages/daemon/src/model/plugin.ts (hunk 3 above)
Comment (FeelyChau) [reply packages/daemon/src/model/plugin.ts:52]: It's the same with function `list`.

**Comment by FeelyChau** [comment]
> I'm -1 on this current implementation which changes the meaning of id in most APIs and it is incorrect.

How about add an option for fuzzy matching, default value is false, it could be compatible.

File: packages/daemon/src/model/plugin.ts (hunk 2 above)
Comment (FeelyChau) [reply packages/daemon/src/model/plugin.ts:60]: It's a prefix of id, not keyword.

Comment (yorkie) [reply packages/daemon/src/model/plugin.ts:60]: I don't think we should introduce the prefix of id in daemon, which would add the complexity of our models.

**Comment by yorkie** [comment]
> > I'm -1 on this current implementation which changes the meaning of id in most APIs and it is incorrect.
> 
> How about add an option for fuzzy matching, default value is false, it could be compatible.

Fuzzy matching doesn't work, because we have defined the `id` in APIs.

**Comment by yorkie** [comment]
As I have commented on https://github.com/alibaba/pipcook/pull/618#discussion_r511729325, the daemon could provide a new API for searching the ids that match, and then the CLI use the result id to request the data.

**Comment by FeelyChau** [comment]
> As I have commented on [#618 (comment)](https://github.com/alibaba/pipcook/pull/618#discussion_r511729325), the daemon could provide a new API for searching the ids that match, and then the CLI use the result id to request the data.

OK, I'll commit though another PR.

**Pull request closed without merging**